File: web/src/util/submissionTags.test.ts
import { describe, expect, it } from "vitest"
import { readFileSync } from "node:fs"
import { fileURLToPath } from "node:url"

import {
  SUBMISSION_TAGS_CAP,
  matchesSubmissionTag,
  parseSubmissionTags,
  safeShimTagPatterns,
  submissionTagsToText,
  validateSubmissionTags,
} from "./submissionTags"

// The web half of the matcher lockstep: the same golden cases the Go
// contract.MatchesSubmissionTag and the Python copies assert. A drift on any
// side fails its fixture test — the pattern strings go verbatim into the
// shim's on.push.tags, so every evaluator must agree on what fires.
describe("matchesSubmissionTag — shared fixture parity", () => {
  const fixtureUrl = new URL(
    "../../../cli/shared/testdata/submission_tag_match_cases.json",
    import.meta.url,
  )
  const doc = JSON.parse(readFileSync(fileURLToPath(fixtureUrl), "utf8")) as {
    cases: { patterns: string[]; tag: string; matches: boolean }[]
  }

  it("has cases", () => {
    expect(doc.cases.length).toBeGreaterThan(0)
  })

  for (const [i, c] of doc.cases.entries()) {
    it(`case ${i}: [${c.patterns.join(", ")}] vs "${c.tag}" -> ${c.matches}`, () => {
      expect(matchesSubmissionTag(c.patterns, c.tag)).toBe(c.matches)
    })
  }
})

describe("matchesSubmissionTag — fail-closed on uncompilable patterns", () => {
  it("an invalid pattern matches nothing and doesn't poison later ones", () => {
    // A reversed character-class range fails RegExp compilation.
    expect(matchesSubmissionTag(["[z-a]"], "m")).toBe(false)
    expect(matchesSubmissionTag(["[z-a]", "good"], "good")).toBe(true)
  })
})

describe("parseSubmissionTags / submissionTagsToText", () => {
  it("splits lines, trims, drops blanks, round-trips", () => {
    expect(parseSubmissionTags(" phase1 \n\nphase2\r\ncomplete\n")).toEqual([
      "phase1",
      "phase2",
      "complete",
    ])
    expect(submissionTagsToText(["phase1", "v*"])).toBe("phase1\nv*")
    expect(submissionTagsToText(undefined)).toBe("")
  })
})

describe("validateSubmissionTags", () => {
  it("accepts valid pattern lists (and the empty list)", () => {
    expect(validateSubmissionTags([])).toBeUndefined()
    expect(
      validateSubmissionTags(["phase1", "v*", "release-[0-9]", "a/b?"]),
    ).toBeUndefined()
  })

  it("rejects exclude patterns, bad charset, duplicates, and over-cap", () => {
    expect(validateSubmissionTags(["!v*"])).toMatch(/exclude/)
    expect(validateSubmissionTags(['ta"g'])).toMatch(/may only use/)
    expect(validateSubmissionTags(["has space"])).toMatch(/may only use/)
    expect(validateSubmissionTags(["dup", "dup"])).toMatch(/more than once/)
    expect(
      validateSubmissionTags(
        Array.from({ length: SUBMISSION_TAGS_CAP + 1 }, (_, i) => `t${i}`),
      ),
    ).toMatch(/Too many/)
  })

  it("rejects stacked/leading quantifiers (Python-divergent patterns)", () => {
    // v*+ etc. compile as POSSESSIVE quantifiers in Python (and match!)
    // while Go/JS throw — the one construct where the four matcher copies
    // would diverge, so writers refuse it (mirrors Go ValidateSubmissionTags).
    for (const p of ["v*+", "a++", "x?+", "m**+", "+lead", "?lead"]) {
      expect(validateSubmissionTags([p])).toMatch(/can't start|follow another/)
    }
  })
})

describe("safeShimTagPatterns — render-time fail-closed gate", () => {
  it("passes safe lists through and fails closed all-or-nothing", () => {
    // Mirrors Go contract.ShimTagsList: the shim writers consume the
    // PUBLISHED (hand-editable) manifest, so one unsafe pattern drops the
    // whole milestone set rather than rendering a partial/hostile tags line.
    expect(safeShimTagPatterns(undefined)).toEqual([])
    expect(safeShimTagPatterns(["phase1", "v*"])).toEqual(["phase1", "v*"])
    expect(safeShimTagPatterns(["v*+"])).toEqual([])
    expect(safeShimTagPatterns(["phase1", 'ta"g'])).toEqual([])
    expect(safeShimTagPatterns(["phase1", "a++"])).toEqual([])
    expect(
      safeShimTagPatterns(
        Array.from({ length: SUBMISSION_TAGS_CAP + 1 }, (_, i) => `t${i}`),
      ),
    ).toEqual([])
  })
})

File: cli/shared/contract/submissiontags_test.go
package contract

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// TestMatchesSubmissionTag_SharedFixtureParity runs the shared golden cases so
// the Go matcher and its web/Python mirrors stay in lockstep — same role the
// control_path_cases.json fixture plays for the allowed_files keep-set.
func TestMatchesSubmissionTag_SharedFixtureParity(t *testing.T) {
	raw, err := os.ReadFile(filepath.Join("..", "testdata", "submission_tag_match_cases.json"))
	if err != nil {
		t.Fatalf("read shared fixture: %v", err)
	}
	var fixture struct {
		Cases []struct {
			Patterns []string `json:"patterns"`
			Tag      string   `json:"tag"`
			Matches  bool     `json:"matches"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatalf("parse shared fixture: %v", err)
	}
	if len(fixture.Cases) == 0 {
		t.Fatal("shared fixture has no cases; did the file move?")
	}
	for _, c := range fixture.Cases {
		if got := MatchesSubmissionTag(c.Patterns, c.Tag); got != c.Matches {
			t.Errorf("MatchesSubmissionTag(%v, %q) = %v, want %v", c.Patterns, c.Tag, got, c.Matches)
		}
	}
}

// TestMatchesSubmissionTag_UncompilablePatternFailsClosed pins the fail-closed
// contract: a pattern the translator can't compile matches nothing (never
// everything) — writer validation should prevent these, but the runner
// re-checks hand-edited manifests.
func TestMatchesSubmissionTag_UncompilablePatternFailsClosed(t *testing.T) {
	// A reversed character-class range ([z-a]) fails regexp compilation.
	if MatchesSubmissionTag([]string{"[z-a]"}, "m") {
		t.Error("uncompilable pattern must match nothing")
	}
	// A bad pattern must not poison a later good one.
	if !MatchesSubmissionTag([]string{"[z-a]", "good"}, "good") {
		t.Error("a later valid pattern must still match")
	}
}

// TestShimTagsList_FailsClosedOnUnsafePatterns pins the render chokepoint:
// the shim writers (gh-student accept, gh-teacher retrofit, and the web
// mirrors) consume the PUBLISHED manifest, so an unsafe pattern that
// bypassed write-time validation must drop the ENTIRE milestone set — the
// rendered tags line falls back to the canonical `"submit/*"` alone rather
// than splicing hostile or divergent content into a student repo's workflow.
func TestShimTagsList_FailsClosedOnUnsafePatterns(t *testing.T) {
	if got := ShimTagsList(nil); got != `"submit/*"` {
		t.Errorf("ShimTagsList(nil) = %q, want the bare canonical entry", got)
	}
	if got := ShimTagsList([]string{"phase1", "v*"}); got != `"phase1", "v*", "submit/*"` {
		t.Errorf("ShimTagsList(safe) = %q", got)
	}
	for _, patterns := range [][]string{
		{"v*+"},           // stacked quantifier (Python-divergent)
		{"phase1", "a++"}, // one bad drops ALL (all-or-nothing)
		{`ta"g`},          // quote would break the YAML string
		{"has space"},     // charset
		{"+lead"},         // leading quantifier
	} {
		if got := ShimTagsList(patterns); got != `"submit/*"` {
			t.Errorf("ShimTagsList(%v) = %q, want fail-closed %q", patterns, got, `"submit/*"`)
		}
	}
	over := make([]string, SubmissionTagsCap+1)
	for i := range over {
		over[i] = "t" + string(rune('a'+i%26))
	}
	// Over-cap also fails closed (dups in `over` are irrelevant here).
	if got := ShimTagsList(over); got != `"submit/*"` {
		t.Errorf("ShimTagsList(over-cap) = %q, want fail-closed", got)
	}
}

// TestMatchesSubmissionTag_StackedQuantifiersFailClosed pins the guard the
// fixture also covers, plus the property that motivates it: these patterns
// must match NOTHING here (and in the web/Python mirrors) even though
// Python's regex dialect would happily compile them possessively.
func TestMatchesSubmissionTag_StackedQuantifiersFailClosed(t *testing.T) {
	for _, pattern := range []string{"v*+", "a++", "x?+", "m**+", "+lead", "?lead"} {
		if MatchesSubmissionTag([]string{pattern}, "v1") {
			t.Errorf("MatchesSubmissionTag([%q], \"v1\") = true, want fail-closed false", pattern)
		}
		if !IsSafeSubmissionTagPattern("phase1") || IsSafeSubmissionTagPattern(pattern) {
			t.Errorf("IsSafeSubmissionTagPattern(%q) should be false (and phase1 true)", pattern)
		}
	}
}
